# 🌦️ Introducción a Python 🐼

## Soluciones a los problemas propuestos



# Ejercicios de aplicacion

## Ejercicio 1: Análisis de Calidad del Agua en un Río

### Tienes datos de concentración de oxígeno disuelto (en mg/L) en diferentes puntos de un río.
### La lista 'oxigeno_disuelto' contiene las mediciones en orden, desde la fuente hasta la desembocadura.

oxigeno_disuelto = [8.5, 8.2, 7.9, 7.6, 7.2, 6.8, 6.5, 6.1, 5.8, 5.5]

- Calcula la concentración promedio de oxígeno disuelto en el río.
- Determina si en algún punto del río la concentración de oxígeno disuelto cae por debajo de 6.0 mg/L, lo cual es crítico para la vida acuática.
- Obtén las mediciones de oxígeno disuelto en el tramo medio del río (desde el cuarto hasta el séptimo punto de medición, inclusive).
- Calcula el porcentaje de disminución de oxígeno disuelto entre el primer y último punto de medición.
### Imprime todos los resultados con mensajes descriptivos.


## Solución: 

### Análisis de Calidad del Agua en un Río


In [ ]:

oxigeno_disuelto = [8.5, 8.2, 7.9, 7.6, 7.2, 6.8, 6.5, 6.1, 5.8, 5.5]


### Cálculo del promedio de oxígeno disuelto

In [ ]:
promedio_oxigeno = sum(oxigeno_disuelto) / len(oxigeno_disuelto)
print(f"1. La concentración promedio de oxígeno disuelto es {promedio_oxigeno:.2f} mg/L")



### Verificar si en algún punto la concentración cae por debajo de 6.0 mg/L


In [ ]:
nivel_critico = any(nivel < 6.0 for nivel in oxigeno_disuelto)
print(f"2. ¿Hay algún punto con nivel crítico de oxígeno disuelto? {nivel_critico}")



### Obtener mediciones del tramo medio del río

In [ ]:
tramo_medio = oxigeno_disuelto[3:7]
print(f"3. Las mediciones en el tramo medio del río son: {tramo_medio}")


### Calcular el porcentaje de disminución entre el primer y último punto


In [ ]:
disminucion_porcentaje = (oxigeno_disuelto[0] - oxigeno_disuelto[-1]) / oxigeno_disuelto[0] * 100
print(f"4. El porcentaje de disminución de oxígeno disuelto es {disminucion_porcentaje:.2f}%")

## Ejercicio 2: Cálculo de volumen de agua
### Un embalse tiene forma rectangular con las siguientes dimensiones:
### largo = 500 m, ancho = 200 m, profundidad = 10 m
### Calcula el volumen de agua en el embalse en metros cúbicos.



In [ ]:
largo = 500
ancho = 200
profundidad = 10

volumen = # Tu código aquí

print(f"El volumen de agua en el embalse es {volumen} metros cúbicos.")



## Ejercicio 3: Comparación de niveles de pH
### El pH neutro es 7. Un pH por debajo de 7 es ácido, por encima es básico.
### Determina si una muestra de agua con pH 6.5 es ácida.


In [ ]:

ph_muestra = 6.5
es_acida = # Tu código aquí

print(f"¿La muestra de agua es ácida? {es_acida}")



## Ejercicio 4: Evaluación de la calidad del agua
### El agua es potable si cumple todas estas condiciones:
- pH entre 6.5 y 8.5
- Turbidez menor a 5 NTU
- Cloro residual entre 0.3 y 1.5 mg/L
### Determina si una muestra de agua es potable con los siguientes valores:


In [ ]:

ph = 7.2
turbidez = 3.5
cloro_residual = 0.5

es_potable = # Tu código aquí

print(f"¿El agua es potable? {es_potable}")



## Ejercicio 5: Análisis de precipitaciones
### Tienes una lista con las precipitaciones diarias (en mm) de la última semana.
### Obtén las precipitaciones de los días laborables (índices 0 a 4).


In [ ]:

precipitaciones = [5.2, 0.0, 12.8, 3.5, 1.0, 22.3, 8.7]
precipitaciones_laborables = # Tu código aquí

print(f"Las precipitaciones en los días laborables fueron: {precipitaciones_laborables}")



## Ejercicio 6: Comparación de caudales
### Tienes dos listas con mediciones de caudal (en m³/s) en dos puntos de un río.
### Determina si el caudal promedio en el segundo punto es mayor que en el primero.


In [ ]:

caudal_punto1 = [23.5, 25.1, 22.8, 24.3, 26.5]
caudal_punto2 = [24.1, 25.3, 23.6, 24.8, 26.2]

promedio_punto1 = # Tu código aquí
promedio_punto2 = # Tu código aquí
caudal_aumenta = # Tu código aquí

print(f"¿El caudal aumenta entre el punto 1 y el punto 2? {caudal_aumenta}")

# Ejercicio de aplicación resuelto

## Problema: Análisis de Calidad del Agua

- Crea una variable llamada 'nombre_rio' y asígnale el valor "Río Colorado".

- Crea una lista llamada 'mediciones_ph' con los siguientes valores de pH: 7.2, 6.8, 7.5, 8.0, 7.1

- Calcula el pH promedio y guárdalo en una variable llamada 'ph_promedio'.

- Crea un mensaje que diga: "El pH promedio del [nombre del río] es [pH promedio]", usando .format() para insertar las variables.

- Convierte el mensaje a mayúsculas usando un método de cadenas.

- Imprime el mensaje final.

# Solución:

- Crear la variable 'nombre_rio'


In [12]:
nombre_rio = "Río Colorado"
nombre_rio

'Río Colorado'


- Crear la lista 'mediciones_ph'


In [7]:
mediciones_ph = [7.2, 6.8, 7.5, 8.0, 7.1]



- Calcular el pH promedio

In [11]:
# Usamos sum() para sumar todos los valores de la lista
# y len() para obtener el número de elementos
ph_promedio = sum(mediciones_ph) / len(mediciones_ph)
ph_promedio


7.32


- Crear el mensaje usando .format()


In [14]:
mensaje = "El pH promedio del {} es {:.2f}".format(nombre_rio, ph_promedio)
mensaje

'El pH promedio del Río Colorado es 7.32'


- Convertir el mensaje a mayúsculas


In [16]:
mensaje_mayusculas = mensaje.upper()
mensaje_mayusculas

'EL PH PROMEDIO DEL RÍO COLORADO ES 7.32'


- Imprimir el mensaje final


In [17]:
print(mensaje_mayusculas)


EL PH PROMEDIO DEL RÍO COLORADO ES 7.32



### El resultado será:
#### EL PH PROMEDIO DEL RÍO COLORADO ES 7.32

### Explicación adicional:
- En el paso 3, usamos sum() y len(), que son funciones integradas de Python.
- En el paso 4, usamos .format() para insertar las variables en el string.
- El {:.2f} indica que queremos mostrar el número con 2 decimales.
- En el paso 5, upper() es un método de cadenas que convierte todo a mayúsculas.

# Ejercicios de aplicación 

## Ejercicio 7: Asignación de variables
- Asigna valores a variables para representar diferentes mediciones en un río.
- Usa nombres de variables apropiados según las reglas mencionadas.
- Desarrolla el código (asigna al menos 3 variables con diferentes tipos de datos)
- Imprime las variables utilizando print()

In [ ]:
# Asignación de variables

# Nombre del río (cadena de caracteres)
nombre_rio = "Río Amazonas"

# Caudal del río en metros cúbicos por segundo (número entero)
caudal_rio = 209000

# Temperatura del agua en grados Celsius (número flotante)
temperatura_agua = 26.5

# Imprimir las variables
print("Nombre del río:", nombre_rio)
print("Caudal del río (m³/s):", caudal_rio)
print("Temperatura del agua (°C):", temperatura_agua)


## Ejercicio 8: Formateo de cadenas
- Utiliza los datos del problema 1 para crear un mensaje formateado sobre el estado del río.
- Usa tanto .format() como los indicadores de formato (%d, %f, %s).
- Desarrolla el código

In [ ]:

# Datos del problema 1
nombre_rio = "Río Amazonas"
caudal_rio = 209000
temperatura_agua = 26.5

# Formateo de cadenas usando .format()
mensaje_format = "El {} tiene un caudal de {} m³/s y una temperatura del agua de {} °C.".format(nombre_rio, caudal_rio, temperatura_agua)
print(mensaje_format)

# Formateo de cadenas usando indicadores de formato (%d, %f, %s)
mensaje_indicadores = "El %s tiene un caudal de %d m³/s y una temperatura del agua de %.1f °C." % (nombre_rio, caudal_rio, temperatura_agua)
print(mensaje_indicadores)


## Ejercicio 9: Manipulación de cadenas
- Crea una cadena que describa los componentes de un sistema hídrico.
- Utiliza al menos 3 métodos de cadenas diferentes (como lower(), upper(), split(), etc.).
- sistema_hidrico = "Río Chucunaque: Cuenca, Afluentes, Desembocadura"
- Desarrolla el código

In [ ]:
## Ejercicio 9: Manipulación de cadenas

# Crear una cadena que describa los componentes de un sistema hídrico
sistema_hidrico = "Río Chucunaque: Cuenca, Afluentes, Desembocadura"

# Utilizar diferentes métodos de cadenas

# Convertir toda la cadena a minúsculas
sistema_hidrico_lower = sistema_hidrico.lower()
print("Cadena en minúsculas:", sistema_hidrico_lower)

# Convertir toda la cadena a mayúsculas
sistema_hidrico_upper = sistema_hidrico.upper()
print("Cadena en mayúsculas:", sistema_hidrico_upper)

# Separar la cadena en componentes usando el delimitador ":"
componentes = sistema_hidrico.split(":")
print("Componentes del sistema hídrico:", componentes)

# Separar los componentes secundarios usando el delimitador ","
componentes_secundarios = componentes[1].split(", ")
print("Componentes secundarios del sistema hídrico:", componentes_secundarios)

# Resultado final
print("Sistema hídrico original:", sistema_hidrico)
print("Sistema hídrico en minúsculas:", sistema_hidrico_lower)
print("Sistema hídrico en mayúsculas:", sistema_hidrico_upper)
print("Componentes principales del sistema hídrico:", componentes)
print("Componentes secundarios del sistema hídrico:", componentes_secundarios)


## Ejercicio 10: Conversiones de tipos de datos
### Convierte los siguientes datos a los tipos especificados:
- El número 15.7 a entero
- El número 24 a float
- El número 30.5 a string

In [ ]:
## Ejercicio 10: Conversiones de tipos de datos

# Datos iniciales
numero_float = 15.7
numero_entero = 24
numero_float_a_string = 30.5

# Convertir el número 15.7 a entero
numero_convertido_a_entero = int(numero_float)
print("El número 15.7 convertido a entero es:", numero_convertido_a_entero)

# Convertir el número 24 a float
numero_convertido_a_float = float(numero_entero)
print("El número 24 convertido a float es:", numero_convertido_a_float)

# Convertir el número 30.5 a string
numero_convertido_a_string = str(numero_float_a_string)
print("El número 30.5 convertido a string es:", numero_convertido_a_string)


## Ejercicio 11: Asignación múltiple y operaciones básicas
- Asigna en una sola línea valores a variables que representen mediciones de pH en diferentes puntos de un río.
- Calcula el pH promedio y muestra el resultado.

In [ ]:
## Ejercicio 11: Asignación múltiple y operaciones básicas

# Asignar en una sola línea valores a variables que representen mediciones de pH en diferentes puntos de un río
ph_punto1, ph_punto2, ph_punto3, ph_punto4 = 7.5, 8.1, 7.8, 7.2

# Calcular el pH promedio
ph_promedio = (ph_punto1 + ph_punto2 + ph_punto3 + ph_punto4) / 4

# Mostrar el resultado
print("El pH promedio en los diferentes puntos del río es:", ph_promedio)


--------------------------------------------------------------------------------------------------------------------------------------------------------

# Ejercicios

## Ejercicio 12: Listas en el Contexto Meteorológico
### Objetivo: Crear y manipular listas de datos meteorológicos.

- Instrucciones:

- Crea una lista llamada temperaturas_diarias que contenga las siguientes temperaturas en grados Celsius: 22.5, 23.3, 21.8, 19.5, 20.1.
- Agrega la temperatura de hoy (por ejemplo, 24.0) a la lista.
- Cambia la segunda temperatura (23.3) por 25.0 debido a una corrección en los datos.
- Elimina la temperatura más baja de la lista.
- Imprime el número de días de los que tienes datos de temperatura.
- Ordena la lista de temperaturas en orden ascendente y luego imprímela.

## Ejercicio 13: Tuplas en el Contexto de Fauna
### Objetivo: Utilizar tuplas para almacenar datos de especies animales y sus características.

- Instrucciones:

- Crea una tupla llamada especie_info que contenga el nombre de una especie animal, su hábitat y su promedio de vida en años. Ejemplo: ("Lince", "Bosque", 15).
- Imprime la información de la especie en un formato claro.
- Accede e imprime solo el hábitat de la especie.

## Ejercicio 14: Diccionarios en el Contexto de Embalses
### Objetivo: Crear y manipular diccionarios que contengan información de embalses.

- Instrucciones:

- Crea un diccionario llamado embalses con la siguiente estructura: 
- embalses = {
    "Embalse A": {"capacidad": 120, "nivel_actual": 80},
    "Embalse B": {"capacidad": 200, "nivel_actual": 150},
    "Embalse C": {"capacidad": 100, "nivel_actual": 70}
  }
- Agrega un nuevo embalse llamado "Embalse D" con una capacidad de 250 y un nivel actual de 200.
- Actualiza el nivel actual del "Embalse A" a 90.
- Elimina el "Embalse C" del diccionario.
- Imprime la capacidad del "Embalse B".

## Ejercicio 15: Conjuntos en el Contexto de Clima
### Objetivo: Utilizar conjuntos para trabajar con datos únicos de fenómenos climáticos.

- Instrucciones:

- Crea un conjunto llamado fenomenos_climaticos que contenga los siguientes fenómenos: "tormenta", "inundación", "sequía".
- Agrega "huracán" al conjunto.
- Intenta agregar "tormenta" nuevamente al conjunto y observa qué sucede.
- Elimina "sequía" del conjunto.
- Imprime todos los fenómenos climáticos en el conjunto.

## Ejercicio 16: Combinando Estructuras de Datos en el Contexto Meteorológico
### Objetivo: Utilizar listas, tuplas, diccionarios y conjuntos en un solo ejercicio para analizar datos meteorológicos.

- Instrucciones:

- Crea una lista de tuplas llamada datos_meteorologicos, donde cada tupla contenga el nombre de una ciudad, la temperatura promedio del día y la condición climática (por ejemplo: [("Madrid", 30, "soleado"), ("Londres", 22, "nublado"), ("Nueva York", 25, "lluvioso")]).
- Convierte esta lista en un diccionario llamado ciudades_meteorologia donde las claves sean los nombres de las ciudades y los valores sean otros diccionarios con las temperaturas y condiciones.
- Crea un conjunto llamado condiciones_unicas que contenga todas las condiciones climáticas únicas presentes en ciudades_meteorologia.
- Imprime las condiciones únicas y la temperatura promedio de Madrid.

# Soluciones

## Ejercicio 12: Listas en el Contexto Meteorológico

**Objetivo**: Crear y manipular listas de datos meteorológicos.

**Instrucciones**:

1. Crea una lista llamada `temperaturas_diarias` que contenga las siguientes temperaturas en grados Celsius: 22.5, 23.3, 21.8, 19.5, 20.1.
2. Agrega la temperatura de hoy (por ejemplo, 24.0) a la lista.
3. Cambia la segunda temperatura (23.3) por 25.0 debido a una corrección en los datos.
4. Elimina la temperatura más baja de la lista.
5. Imprime el número de días de los que tienes datos de temperatura.
6. Ordena la lista de temperaturas en orden ascendente y luego imprímela.

```python
# Solución
temperaturas_diarias = [22.5, 23.3, 21.8, 19.5, 20.1]
temperaturas_diarias.append(24.0)
temperaturas_diarias[1] = 25.0
temperaturas_diarias.remove(min(temperaturas_diarias))
print("Número de días con datos de temperatura:", len(temperaturas_diarias))
temperaturas_diarias.sort()
print("Temperaturas ordenadas:", temperaturas_diarias)



---------

# Ejercicio 13: Tuplas en el Contexto de Fauna

**Objetivo**: Utilizar tuplas para almacenar datos de especies animales y sus características.

**Instrucciones**:

1. Crea una tupla llamada `especie_info` que contenga el nombre de una especie animal, su hábitat y su promedio de vida en años. Ejemplo: ("Lince", "Bosque", 15).
2. Imprime la información de la especie en un formato claro.
3. Accede e imprime solo el hábitat de la especie.

```python
# Solución
especie_info = ("Lince", "Bosque", 15)
print(f"Especie: {especie_info[0]}, Hábitat: {especie_info[1]}, Promedio de vida: {especie_info[2]} años")
print("Hábitat de la especie:", especie_info[1])




---------

# Ejercicio 14: Diccionarios en el Contexto de Embalses

**Objetivo**: Crear y manipular diccionarios que contengan información de embalses.

**Instrucciones**:

1. Crea un diccionario llamado `embalses` con la siguiente estructura:
   ```python
   embalses = {
       "Embalse A": {"capacidad": 120, "nivel_actual": 80},
       "Embalse B": {"capacidad": 200, "nivel_actual": 150},
       "Embalse C": {"capacidad": 100, "nivel_actual": 70}
   }
Agrega un nuevo embalse llamado "Embalse D" con una capacidad de 250 y un nivel actual de 200.
Actualiza el nivel actual del "Embalse A" a 90.
Elimina el "Embalse C" del diccionario.
Imprime la capacidad del "Embalse B".

```python
# Solución
embalses = {
    "Embalse A": {"capacidad": 120, "nivel_actual": 80},
    "Embalse B": {"capacidad": 200, "nivel_actual": 150},
    "Embalse C": {"capacidad": 100, "nivel_actual": 70}
}

embalses["Embalse D"] = {"capacidad": 250, "nivel_actual": 200}
embalses["Embalse A"]["nivel_actual"] = 90
del embalses["Embalse C"]

print("Capacidad del Embalse B:", embalses["Embalse B"]["capacidad"])


---------

# Ejercicio 15: Conjuntos en el Contexto de Clima

**Objetivo**: Utilizar conjuntos para trabajar con datos únicos de fenómenos climáticos.

**Instrucciones**:

1. Crea un conjunto llamado `fenomenos_climaticos` que contenga los siguientes fenómenos: "tormenta", "inundación", "sequía".
2. Agrega "huracán" al conjunto.
3. Intenta agregar "tormenta" nuevamente al conjunto y observa qué sucede.
4. Elimina "sequía" del conjunto.
5. Imprime todos los fenómenos climáticos en el conjunto.

```python
# Solución
fenomenos_climaticos = {"tormenta", "inundación", "sequía"}
fenomenos_climaticos.add("huracán")
fenomenos_climaticos.add("tormenta")  # No se agregará porque ya existe en el conjunto
fenomenos_climaticos.remove("sequía")
print("Fenómenos climáticos:", fenomenos_climaticos)

---------

# Ejercicio 16: Combinando Estructuras de Datos en el Contexto Meteorológico

**Objetivo**: Utilizar listas, tuplas, diccionarios y conjuntos en un solo ejercicio para analizar datos meteorológicos.

**Instrucciones**:

1. Crea una lista de tuplas llamada `datos_meteorologicos`, donde cada tupla contenga el nombre de una ciudad, la temperatura promedio del día y la condición climática (por ejemplo: [("Madrid", 30, "soleado"), ("Londres", 22, "nublado"), ("Nueva York", 25, "lluvioso")]).
2. Convierte esta lista en un diccionario llamado `ciudades_meteorologia` donde las claves sean los nombres de las ciudades y los valores sean otros diccionarios con las temperaturas y condiciones.
3. Crea un conjunto llamado `condiciones_unicas` que contenga todas las condiciones climáticas únicas presentes en `ciudades_meteorologia`.
4. Imprime las condiciones únicas y la temperatura promedio de Madrid.

```python
# Solución
datos_meteorologicos = [("Madrid", 30, "soleado"), ("Londres", 22, "nublado"), ("Nueva York", 25, "lluvioso")]

ciudades_meteorologia = {ciudad: {"temperatura": temp, "condicion": cond} for ciudad, temp, cond in datos_meteorologicos}

condiciones_unicas = {info["condicion"] for info in ciudades_meteorologia.values()}

print("Condiciones climáticas únicas:", condiciones_unicas)
print("Temperatura promedio en Madrid:", ciudades_meteorologia["Madrid"]["temperatura"])

---------

# Ejercicio 17: Tuplas en el Contexto de Fauna

**Objetivo**: Utilizar tuplas para almacenar datos de especies animales y sus características.

**Instrucciones**:

1. Crea una tupla llamada `especie_info` que contenga el nombre de una especie animal, su hábitat y su promedio de vida en años. Ejemplo: ("Lince", "Bosque", 15).
2. Imprime la información de la especie en un formato claro.
3. Accede e imprime solo el hábitat de la especie.

```python
# Solución
especie_info = ("Lince", "Bosque", 15)
print(f"Especie: {especie_info[0]}, Hábitat: {especie_info[1]}, Promedio de vida: {especie_info[2]} años")
print("Hábitat de la especie:", especie_info[1])


---------

# Ejercicio 18: Diccionarios en el Contexto de Embalses

**Objetivo**: Crear y manipular diccionarios que contengan información de embalses.

**Instrucciones**:

1. Crea un diccionario llamado `embalses` con la siguiente estructura:
   ```python
   embalses = {
       "Embalse A": {"capacidad": 120, "nivel_actual": 80},
       "Embalse B": {"capacidad": 200, "nivel_actual": 150},
       "Embalse C": {"capacidad": 100, "nivel_actual": 70}
   }
Agrega un nuevo embalse llamado "Embalse D" con una capacidad de 250 y un nivel actual de 200.
Actualiza el nivel actual del "Embalse A" a 90.
Elimina el "Embalse C" del diccionario.
Imprime la capacidad del "Embalse B".

```python
# Solución
embalses = {
    "Embalse A": {"capacidad": 120, "nivel_actual": 80},
    "Embalse B": {"capacidad": 200, "nivel_actual": 150},
    "Embalse C": {"capacidad": 100, "nivel_actual": 70}
}

embalses["Embalse D"] = {"capacidad": 250, "nivel_actual": 200}
embalses["Embalse A"]["nivel_actual"] = 90
del embalses["Embalse C"]

print("Capacidad del Embalse B:", embalses["Embalse B"]["capacidad"])

--------------------------------------------------------------------------------------------------------------------------------------------------------

# Ejercicio 19: Listas en el Contexto de Especies Vegetales

#### Objetivo: Crear y manipular listas de datos sobre especies vegetales en Panamá.
Instrucciones:

1. Crea una lista llamada `especies_vegetales` que contenga los siguientes nombres de especies: "Ceiba", "Guayacán", "Roble", "Caoba", "Cedro".

2. Agrega una nueva especie (por ejemplo, "Mango") a la lista.

3. Cambia el segundo nombre de especie ("Guayacán") por "Caimito" debido a una corrección en los datos.

4. Elimina la especie con el nombre más corto de la lista.

5. Imprime el número de especies de las que tienes datos.

6. Ordena la lista de especies en orden alfabético y luego imprímela.

In [3]:
# Crear la lista de especies vegetales
especies_vegetales = ["Ceiba", "Guayacán", "Roble", "Caoba", "Cedro"]

# Agregar una nueva especie
especies_vegetales.append("Mango")

# Cambiar la segunda especie ("Guayacán") por "Caimito"
especies_vegetales[1] = "Caimito"

# Eliminar la especie con el nombre más corto de la lista
# Para ello, buscamos el elemento con el nombre más corto
especie_mas_corta = min(especies_vegetales, key=len)
especies_vegetales.remove(especie_mas_corta)

# Imprimir el número de especies de las que tenemos datos
print(f"El número de especies es: {len(especies_vegetales)}")

# Ordenar la lista de especies en orden alfabético y luego imprimirla
especies_vegetales.sort()
print("Lista de especies ordenada alfabéticamente:")
print(especies_vegetales)

El número de especies es: 5
Lista de especies ordenada alfabéticamente:
['Caimito', 'Caoba', 'Cedro', 'Mango', 'Roble']


# Ejercicio 20: Conjuntos en el Contexto de Clima

**Objetivo**: Utilizar conjuntos para trabajar con datos únicos de fenómenos climáticos.

**Instrucciones**:

1. Crea un conjunto llamado `fenomenos_climaticos` que contenga los siguientes fenómenos: "tormenta", "inundación", "sequía".
2. Agrega "huracán" al conjunto.
3. Intenta agregar "tormenta" nuevamente al conjunto y observa qué sucede.
4. Elimina "sequía" del conjunto.
5. Imprime todos los fenómenos climáticos en el conjunto.

```python
# Solución
fenomenos_climaticos = {"tormenta", "inundación", "sequía"}
fenomenos_climaticos.add("huracán")
fenomenos_climaticos.add("tormenta")  # No se agregará porque ya existe en el conjunto
fenomenos_climaticos.remove("sequía")
print("Fenómenos climáticos:", fenomenos_climaticos)

# Ejercicio de aplicación resuelto

In [79]:
# Problema: Análisis de Calidad del Agua en Embalses

# Tienes datos de calidad del agua de varios embalses. Cada embalse tiene mediciones 
# de diferentes parámetros tomadas en distintos días. Tu tarea es analizar estos datos 
# y proporcionar un informe.

datos_embalses = {
    "Embalse A": [
        ("2023-06-01", {"pH": 7.2, "temperatura": 22.5, "turbidez": 3.7}),
        ("2023-06-02", {"pH": 7.3, "temperatura": 23.1, "turbidez": 3.5}),
        ("2023-06-03", {"pH": 7.1, "temperatura": 22.8, "turbidez": 4.0})
    ],
    "Embalse B": [
        ("2023-06-01", {"pH": 6.9, "temperatura": 21.8, "turbidez": 4.2}),
        ("2023-06-02", {"pH": 7.0, "temperatura": 22.2, "turbidez": 4.0}),
        ("2023-06-03", {"pH": 6.8, "temperatura": 22.5, "turbidez": 4.5})
    ],
    "Embalse C": [
        ("2023-06-01", {"pH": 7.5, "temperatura": 23.5, "turbidez": 3.2}),
        ("2023-06-02", {"pH": 7.4, "temperatura": 24.1, "turbidez": 3.3}),
        ("2023-06-03", {"pH": 7.6, "temperatura": 23.8, "turbidez": 3.1})
    ]
}

# Tareas:
# 1. Calcula el promedio de pH, temperatura y turbidez para cada embalse.
# 2. Identifica el embalse con la temperatura más alta y en qué fecha ocurrió.
# 3. Clasifica cada embalse según su pH promedio: 
#    "Ácido" si pH < 7, "Neutral" si 7 <= pH <= 7.5, "Alcalino" si pH > 7.5
# 4. Crea un conjunto con los embalses que tienen al menos una medición de turbidez superior a 4.0.
# 5. Imprime un informe con todos estos resultados.



In [82]:
# Solución: Análisis de Calidad del Agua en Embalses

# 1. Calcular promedios
promedios = {}
for embalse, mediciones in datos_embalses.items():
    suma_pH = suma_temp = suma_turb = 0
    for _, datos in mediciones:
        suma_pH += datos["pH"]
        suma_temp += datos["temperatura"]
        suma_turb += datos["turbidez"]
    n = len(mediciones)
    promedios[embalse] = {
        "pH": suma_pH / n,
        "temperatura": suma_temp / n,
        "turbidez": suma_turb / n
    }
print("Promedios de Calidad del Agua por Embalse:")
print("------------------------------------------")

for embalse, datos in promedios.items():
    print(f"\nEmbalse: {embalse}")
    print(f"  pH promedio:          {datos['pH']:.2f}")
    print(f"  Temperatura promedio: {datos['temperatura']:.2f}°C")
    print(f"  Turbidez promedio:    {datos['turbidez']:.2f} NTU")

print("\n------------------------------------------")

# Explicación:
# 1. Usamos un bucle for para iterar sobre el diccionario 'promedios'.
# 2. Para cada embalse, imprimimos su nombre y los promedios de cada parámetro.
# 3. Utilizamos f-strings para formatear la salida, incluyendo el nombre del embalse y los valores.
# 4. Los valores se formatean a dos decimales usando :.2f para mejor legibilidad.
# 5. Añadimos unidades a la temperatura (°C) y turbidez (NTU) para claridad.
# 6. Usamos sangrías y líneas separadoras para mejorar la presentación visual.


Promedios de Calidad del Agua por Embalse:
------------------------------------------

Embalse: Embalse A
  pH promedio:          7.20
  Temperatura promedio: 22.80°C
  Turbidez promedio:    3.73 NTU

Embalse: Embalse B
  pH promedio:          6.90
  Temperatura promedio: 22.17°C
  Turbidez promedio:    4.23 NTU

Embalse: Embalse C
  pH promedio:          7.50
  Temperatura promedio: 23.80°C
  Turbidez promedio:    3.20 NTU

------------------------------------------


In [ ]:

# 3. Clasificar embalses por pH
clasificacion_pH = {}
for embalse, prom in promedios.items():
    if prom["pH"] < 7:
        clasificacion_pH[embalse] = "Ácido"
    elif prom["pH"] <= 7.5:
        clasificacion_pH[embalse] = "Neutral"
    else:
        clasificacion_pH[embalse] = "Alcalino"


In [ ]:

# 4. Embalses con turbidez > 4.0
embalses_alta_turbidez = set()
for embalse, mediciones in datos_embalses.items():
    for _, datos in mediciones:
        if datos["turbidez"] > 4.0:
            embalses_alta_turbidez.add(embalse)
            break


In [ ]:

# 5. Imprimir informe
print("Informe de Calidad del Agua en Embalses\n")

print("1. Promedios por embalse:")
for embalse, prom in promedios.items():
    print(f"{embalse}:")
    print(f"  pH: {prom['pH']:.2f}")
    print(f"  Temperatura: {prom['temperatura']:.2f}°C")
    print(f"  Turbidez: {prom['turbidez']:.2f} NTU")

print(f"\n2. Temperatura más alta: {temp_max}°C")
print(f"   Embalse: {embalse_max}")
print(f"   Fecha: {fecha_max}")

print("\n3. Clasificación por pH:")
for embalse, clas in clasificacion_pH.items():
    print(f"{embalse}: {clas}")

print("\n4. Embalses con alta turbidez (>4.0 NTU):")
print(", ".join(embalses_alta_turbidez) if embalses_alta_turbidez else "Ninguno")

# Ejercicios de aplicación

In [1]:
# Problema 1: Análisis de Precipitaciones
# Crea una lista con las precipitaciones diarias (en mm) de un mes.
# Utiliza un bucle for para calcular la precipitación total y el promedio.
# Usa una estructura if para clasificar el mes como "seco" si el promedio es menor a 50 mm,
# "normal" si está entre 50 y 100 mm, y "lluvioso" si es mayor a 100 mm.

precipitaciones = [12.5, 8.0, 0.0, 22.3, 15.7, 30.1, 0.0, 5.2, 18.9, 7.6, 
                   0.0, 0.0, 2.1, 9.8, 25.4, 11.2, 0.0, 3.5, 14.7, 6.9, 
                   0.0, 1.8, 4.3, 8.7, 20.1, 13.5, 0.0, 6.4, 17.2, 9.1]

# Inicializar la suma de precipitaciones
precipitacion_total = 0

# Utilizar un bucle for para calcular la precipitación total
for precipitacion in precipitaciones:
    precipitacion_total += precipitacion

# Calcular el promedio de precipitaciones
promedio_precipitacion = precipitacion_total / len(precipitaciones)

# Clasificar el mes según el promedio de precipitaciones
if promedio_precipitacion < 50:
    clasificacion_mes = "seco"
elif 50 <= promedio_precipitacion <= 100:
    clasificacion_mes = "normal"
else:
    clasificacion_mes = "lluvioso"

# Mostrar los resultados
print("Precipitación total del mes:", precipitacion_total, "mm")
print("Promedio diario de precipitación:", promedio_precipitacion, "mm")
print("Clasificación del mes:", clasificacion_mes)


Precipitación total del mes: 275.0 mm
Promedio diario de precipitación: 9.166666666666666 mm
Clasificación del mes: seco


In [2]:
# Problema 2: Registro de Especies
# Crea un diccionario donde las claves sean nombres de especies y los valores sean 
# el número de avistamientos. Usa un bucle for para imprimir cada especie y su conteo.
# Luego, encuentra la especie con más avistamientos.

avistamientos = {
    "águila real": 5,
    "oso pardo": 2,
    "lobo ibérico": 3,
    "lince ibérico": 1,
    "buitre leonado": 8,
    "cabra montés": 12
}

# Utilizar un bucle for para imprimir cada especie y su conteo
print("Avistamientos de especies:")
for especie, conteo in avistamientos.items():
    print(f"{especie}: {conteo} avistamientos")

# Encontrar la especie con más avistamientos
especie_mas_avistamientos = max(avistamientos, key=avistamientos.get)

# Mostrar la especie con más avistamientos
print(f"\nLa especie con más avistamientos es: {especie_mas_avistamientos} con {avistamientos[especie_mas_avistamientos]} avistamientos")


Avistamientos de especies:
águila real: 5 avistamientos
oso pardo: 2 avistamientos
lobo ibérico: 3 avistamientos
lince ibérico: 1 avistamientos
buitre leonado: 8 avistamientos
cabra montés: 12 avistamientos

La especie con más avistamientos es: cabra montés con 12 avistamientos


In [3]:
# Problema 3: Análisis de Calidad del Agua
# Crea una lista de tuplas, donde cada tupla contenga el nombre de un río y su pH.
# Usa un bucle for con tuple unpacking para clasificar cada río como 
# "ácido" (pH < 7), "neutral" (pH == 7), o "alcalino" (pH > 7).

rios_ph = [("Tajo", 7.1), ("Ebro", 8.2), ("Duero", 6.8), 
           ("Guadalquivir", 7.8), ("Guadiana", 6.9)]

# Utilizar un bucle for con tuple unpacking para clasificar cada río
for rio, ph in rios_ph:
    if ph < 7:
        clasificacion = "ácido"
    elif ph == 7:
        clasificacion = "neutral"
    else:
        clasificacion = "alcalino"
    
    print(f"El río {rio} tiene un pH de {ph} y es clasificado como {clasificacion}.")


El río Tajo tiene un pH de 7.1 y es clasificado como alcalino.
El río Ebro tiene un pH de 8.2 y es clasificado como alcalino.
El río Duero tiene un pH de 6.8 y es clasificado como ácido.
El río Guadalquivir tiene un pH de 7.8 y es clasificado como alcalino.
El río Guadiana tiene un pH de 6.9 y es clasificado como ácido.


In [4]:
# Problema 4: Estaciones Meteorológicas
# Crea un conjunto con los nombres de las estaciones que reportaron datos hoy,
# y otro conjunto con las que se espera que reporten mañana.
# Usa operaciones de conjuntos para encontrar:
# 1. Estaciones que reportarán ambos días
# 2. Estaciones que reportarán mañana pero no reportaron hoy
# 3. Estaciones que reportaron hoy pero no reportarán mañana

estaciones_hoy = {"Madrid", "Barcelona", "Sevilla", "Valencia", "Zaragoza"}
estaciones_manana = {"Madrid", "Barcelona", "Sevilla", "Bilbao", "Málaga"}

# 1. Estaciones que reportarán ambos días
estaciones_ambos_dias = estaciones_hoy.intersection(estaciones_manana)
print("Estaciones que reportarán ambos días:", estaciones_ambos_dias)

# 2. Estaciones que reportarán mañana pero no reportaron hoy
estaciones_solo_manana = estaciones_manana.difference(estaciones_hoy)
print("Estaciones que reportarán mañana pero no reportaron hoy:", estaciones_solo_manana)

# 3. Estaciones que reportaron hoy pero no reportarán mañana
estaciones_solo_hoy = estaciones_hoy.difference(estaciones_manana)
print("Estaciones que reportaron hoy pero no reportarán mañana:", estaciones_solo_hoy)


Estaciones que reportarán ambos días: {'Barcelona', 'Sevilla', 'Madrid'}
Estaciones que reportarán mañana pero no reportaron hoy: {'Bilbao', 'Málaga'}
Estaciones que reportaron hoy pero no reportarán mañana: {'Valencia', 'Zaragoza'}


In [5]:
# Problema 5: Análisis de Caudales
# Crea un diccionario donde las claves sean nombres de ríos y los valores sean 
# listas con los caudales diarios (m³/s) durante una semana.
# Usa bucles anidados para encontrar el río con el caudal máximo y el día en que ocurrió.

caudales = {
    "Ebro": [600, 580, 620, 590, 605, 615, 595],
    "Tajo": [320, 310, 330, 305, 325, 315, 318],
    "Duero": [400, 420, 380, 410, 405, 415, 390],
    "Guadalquivir": [280, 290, 270, 285, 275, 295, 288]
}

# Inicializar variables para rastrear el caudal máximo y su correspondiente río y día
caudal_maximo = 0
rio_caudal_maximo = ""
dia_caudal_maximo = 0

# Utilizar bucles anidados para encontrar el río con el caudal máximo y el día en que ocurrió
for rio, caudales_semanales in caudales.items():
    for dia, caudal in enumerate(caudales_semanales):
        if caudal > caudal_maximo:
            caudal_maximo = caudal
            rio_caudal_maximo = rio
            dia_caudal_maximo = dia + 1  # +1 para ajustar a días de la semana (1-7)

# Mostrar el resultado
print(f"El río con el caudal máximo es el {rio_caudal_maximo} con {caudal_maximo} m³/s, ocurrido el día {dia_caudal_maximo} de la semana.")

El río con el caudal máximo es el Ebro con 620 m³/s, ocurrido el día 3 de la semana.
